engine/crates/operation-checks/src/schema/wrapper_types.rs

```diff
@@ -0,0 +1,107 @@
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(u8)]
+pub enum WrapperType {
+    List = 0b10,
+    Required = 0b01,
+    RequiredList = 0b11,
+}
+
+impl WrapperType {
+    const LIST: u8 = WrapperType::List as u8;
+    const REQUIRED: u8 = WrapperType::Required as u8;
+    const REQUIRED_LIST: u8 = WrapperType::RequiredList as u8;
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum WrapperTypes {
+    /// A compact representation for up to four levels of wrapper types. Each pair of bits
+    /// corresponds to an Option<WrapperType> (0b00 = None, 0b01 = Some(Required), 0b10 =>
+    /// Some(List), 0b11 => Some(RequiredList)). The least significant pair of bits corresponds to
+    /// the outermost wrappers.
+    Small(u8),
+    /// This variant is for extreme cases of list fields with more than three levels of nested
+    /// lists. From outermost to innermost.
+    Large(Box<[WrapperType]>),
```

Comment:
Just a thought, but would it be worth making this a `struct WrapperTypes(u32)` and always doing bitwise encoding?

Gives you space for 16 lists and still uses less memory than this will

Reply:
Yes even with this type, there's more room than the u8 if we want. Even though lists nested 16 level deep isn't a realistic use case, I had some worries about not supporting it. I'll open a PR when there's some downtime with the change mentioned in my last comment and this, that should simplify things a lot.

Reply:
Actually, since I'm waiting on other PRs now, let's try right away.

Reply:
0e5649e1f37dccb4027011a2055543bec6a4779f reimplements the same API but with a single integer for the wrappers. I'd like to improve the API though, doing this now.